providers/fips/fipsprov.c

```diff
@@ -667,11 +669,11 @@ int OSSL_provider_init(const OSSL_PROVIDER *provider,
         goto err;
 
     *out = fips_dispatch_table;
-    *provctx = libctx;
 
     return 1;
  err:
     fips_teardown(*provctx);
+    *provctx = NULL;
```

Comment:
I would actually perfer it to use libctx here if it can?

Reply:
I don't understand.  if the init function fails, it should clean up the provider context it created, shouldn't it?

Reply:
I was thinking that the last thing it should do it assign it to the provctx just before it returns when it passes. On the error case I dont like it setting up the pointer so much..

Reply:
... which means that when we have a proper provider CTX that's its own structure that *contains* a libctx, we will have two teardown procedures, one "special" in this init function, and one that's a downcall.  In other words, code copying all over again, with all the problems that may follow.

No thank you.

Reply:
Fair enough - but what we want to make sure is that NO information leaks back to the core if the self test fails. If it does then that is a compliance failure. (That is all I am concerned about with this).

Reply:
I completely understand that concern.  Are you trying to say that attaching `libctx` to `*provctx` early is a risk of such leakage?  That has me thinking of these lines:

https://github.com/openssl/openssl/blob/3f2a8d971a44b9aac30a5725ae44cfcb1d4156f1/crypto/provider_core.c#L489-L491

As one can see, the address to `prov->provctx` is passed directly.  We could easily change that code to use an intermediary temporary variable, this only attaching the provider context to the provider after the init function has returned.  Will that be enough to clear your concerns?  I would rather do that.

Reply:
Its probably ok with what you have now..

Reply:
I just saw your teardown is a one line function which is not really much of a duplication of code.. But anyways..

Reply:
Well, I added such a commit anyway.  Others may be equally concerned, so it's just as well that we try to lessen that concern in a general fashion.